# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [7]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\akgoel\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [8]:
# Read your api key environment variable
load_dotenv()
api_key = os.getenv("news_api")

In [9]:
# Create a newsapi client
from newsapi import NewsApiClient
newsapi = NewsApiClient(api_key=api_key)

In [17]:
# Fetch the most relevant 100 Bitcoin news articles for the last 2 weeks from 5/25-6/8 2021
btc_articles = newsapi.get_everything(
    q="bitcoin",
    language="en",
    sort_by="relevancy",
    from_param="2021-05-25",
    page_size=100
)

In [20]:
# Fetch the most relevant 100 Ethereum news articles for the last 2 weeks from 5/25-6/8 2021
eth_articles = newsapi.get_everything(
    q="ethereum",
    language="en",
    sort_by="relevancy",
    from_param="2021-05-25",
    page_size=100
)

In [22]:
eth_articles["articles"][0]

{'source': {'id': 'reuters', 'name': 'Reuters'},
 'author': 'Reuters',
 'title': 'Ethereum extends gains to rise 8%; bitcoin firms - Reuters',
 'description': 'Cryptocurrency Ethereum extended gains to rise more than 8% on Monday to $2,587 but remained 40% below a record high of above $4,300 hit earlier this month.',
 'url': 'https://www.reuters.com/technology/ethereum-rises-5-2509-bitcoin-firm-2021-05-31/',
 'urlToImage': 'https://www.reuters.com/resizer/yeaYVdBQ4_EGCbJU5Z5E6qExbDw=/1200x628/smart/filters:quality(80)/cloudfront-us-east-2.images.arcpublishing.com/reuters/VMU2AOPUXRPU7KWLBR6GOWQHCM.jpg',
 'publishedAt': '2021-05-31T12:17:00Z',
 'content': 'A representation of virtual currency Ethereum is seen in front of a stock graph in this illustration taken February 19, 2021. REUTERS/Dado Ruvic/Illustration/File PhotoCryptocurrency Ethereum extende… [+1099 chars]'}

In [25]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiments = []
for article in btc_articles["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]        
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        btc_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
        })
    except AttributeError:
        pass

btc_df = pd.DataFrame(btc_sentiments)

text        date  compound  \
0   Hello friends, and welcome back to Week in Rev...  2021-06-05    0.7500   
1   In spite of the environmental and regulatory i...  2021-05-27   -0.5267   
2   By Mary-Ann RussonBusiness reporter, BBC News\...  2021-06-08    0.3400   
3   The ban affects licensed and unlicensed miners...  2021-05-26   -0.5574   
4   After years of hesitation, PayPal continues to...  2021-05-27   -0.2732   
..                                                ...         ...       ...   
95  Dinh said he initially reached out to Bankman-...  2021-06-04    0.8481   
96  Opinions expressed by Entrepreneur contributor...  2021-05-26    0.5267   
97  (Reuters) - Shares of GameStop soared 16.3% on...  2021-05-25    0.2960   
98  By Reuters Staff\r\nFILE PHOTO: Representation...  2021-06-03    0.0000   
99  U.S. one dollar banknotes are seen in front of...  2021-05-25    0.0000   

    positive  negative  neutral  
0      0.154     0.000    0.846  
1      0.000     0.096    0.904  
2      0.070     0.000    0.930  
3      0.000     0.107    0.893  
4      0.000     0.062    0.938  
..       ...       ...      ...  
95     0.223     0.000    0.777  
96     0.155     0.000    0.845  
97     0.064     0.000    0.936  
98     0.000     0.000    1.000  
99     0.000     0.000    1.000  

[100 rows x 6 columns]

In [26]:
# Create the Ethereum sentiment scores DataFrame
eth_sentiments = []
for article in eth_articles["articles"]:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]        
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        eth_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
        })
    except AttributeError:
        pass

eth_df = pd.DataFrame(eth_sentiments)

In [28]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound   positive    negative     neutral
count  100.000000  100.00000  100.000000  100.000000
mean     0.128253    0.05951    0.027570    0.912930
std      0.368170    0.06921    0.049981    0.085346
min     -0.750600    0.00000    0.000000    0.582000
25%      0.000000    0.00000    0.000000    0.853000
50%      0.000000    0.05150    0.000000    0.926000
75%      0.386825    0.09825    0.057500    1.000000
max      0.921700    0.35100    0.262000    1.000000

In [29]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound    positive   negative     neutral
count  100.000000  100.000000  100.00000  100.000000
mean     0.224866    0.083490    0.02820    0.888300
std      0.393903    0.068916    0.05300    0.081151
min     -0.757900    0.000000    0.00000    0.655000
25%      0.000000    0.000000    0.00000    0.844250
50%      0.282100    0.074500    0.00000    0.899500
75%      0.536525    0.125750    0.04275    0.938000
max      0.858100    0.252000    0.26200    1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: Ethereum had the highest mean positive score with 0.08349 compared to Bitcoin at 0.05951

Q: Which coin had the highest compound score?

A: Ethereum had the highest mean compound score at 0.224866 compared to Bitcoin at 0.128253. However, Bitcoin had the highest/max compound score at 0.9127, compared to Ethereum at 0.8581.

Q. Which coin had the highest positive score?

A: Ethereum had the hightest mean positive score of 0.08349 compared to Bitcoin at 0.05951. However, Bitcoin had the highest/max positive score of 0.351 compared to Ethereum at 0.252.

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [10]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [11]:
# Instantiate the lemmatizer
# YOUR CODE HERE!

# Create a list of stopwords
# YOUR CODE HERE!

# Expand the default stopwords list if necessary
# YOUR CODE HERE!

In [12]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text

   
    # Create a tokenized list of the words
    
    
    # Lemmatize words into root words

   
    # Convert the words to lowercase
    
    
    # Remove the stop words
    
    
    return tokens

In [13]:
# Create a new tokens column for Bitcoin
# YOUR CODE HERE!

In [14]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---